### PR TITLE
Cm/change to user driver

### PR DESCRIPTION
**Pull request by CheezItMan**
A 1st pass at the README for the OO-Ride-Share.

Pulling Driver out as a user-created class forces us to either have the user write some `TripDispatcher` methods for loading the data from a CSV or we create the methods and they write Driver to match.  Currently I have them writing the methods just to enable them to practice instantiating objects. 

Right now Wave 1 involves adding small methods to `User` to get used to the codebase.

Wave 2 involves adding the Driver class and making additions to `Trip` and `TripDispatcher` to make use of the new class.

Wave 3 is back to the older code writing methods to add trips and select drivers, and Wave 4 has the enhanced driver selection algorithm.  It seems like a lot, but it **is** a pair project and a 5-day week.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -8,101 +8,79 @@ This is a [level 2](https://github.com/Ada-Developers-Academy/pedagogy/blob/mast
...
 -----|-----
-driver|retrieve the associated driver instance
-passenger|retrieve the associated passenger instance
+User|retrieve the associated User instance
 
 #### TripDispatcher
 The `TripDispatcher` has:
 
-**Attribute**|**Description**
+**Attribute**|**Description**|Returns
```
Comment (tildeee) [inline README.md]: this table doesn't get formatted correctly

File: README.md [hunk 2]
```diff
@@ -8,101 +8,79 @@ This is a [level 2](https://github.com/Ada-Developers-Academy/pedagogy/blob/mast
...
--   `Driver`s
--   `Passenger`s
+-   `User`s
 -   `Trip`s
 
-All of this data is managed in a class called `TripDispatcher`. Our program will contain _one_ instance of `TripDispatcher`, which will load and manage the lists of `Driver`s, `Passenger`s and `Trip`s.
+All of this data is managed in a class called `TripDispatcher`. Our program will contain _one_ instance of `TripDispatcher`, which will load and manage the lists of `Driver`s, `User`s and `Trip`s.
 
-We are going to continue making functionality that works with this data, such as finding the duration of a specific trip or the total amount of money a passenger has spent, and also make functionality to create a new trip.
+We are going to continue making functionality that works with this data, such as adding finding the duration of a specific trip or the total amount of money a User has spent, or the amount of money a driver has made and also make functionality to create a new trip.
```
Comment (tildeee) [inline README.md]: grammar: "such as adding finding the duration"

File: README.md [hunk 3]
```diff
@@ -156,25 +134,71 @@ Currently our implementation saves the start and end time of each trip as a stri
...
+driver|The `Driver` for the trip
+
+Each `Trip` instance should also be able to do the following:
+
+**Method**|**Description**
+-----|-----
+driver|retrieve the associated `Driver` instance
+
+#### Loading Drivers
+Update the `TripDispatcher` class to add the following Methods:
```
Comment (tildeee) [inline README.md]: Do students also need to update the existing `TripDispatcher` functionality that `Trip`s loaded in must now have a driver associated with it?

**Comment by CheezItMan** [comment]
@tildeee Thanks for the feedback.  I was going to look at either:

1.  Providing commented out code for TripDispatcher
1.  Providing more tests or scaffolded tests for Wave 2
1.  Making Wave 4 Optional

And judging based on how much time I think it'll take.

File: README.md [hunk 4]
```diff
@@ -8,101 +8,79 @@ This is a [level 2](https://github.com/Ada-Developers-Academy/pedagogy/blob/mast
...
-
-Each `Passenger` instance is able to:
-
-**Method**|**Description**
------|-----
-get_drivers|retrieve the list of all previous driver instances associated with trips this passenger has taken
+id|The User's ID number
+name|The name of the User
+Phone Number|The User's Phone Number which must be in phone number format (XXX) XXX-XXXX
+trips|A list of trips that only this User has taken as a passenger
```
Comment (droberts-ada) [inline README.md]: Line 37: do you need the `only`?

File: README.md [hunk 5]
```diff
@@ -8,101 +8,79 @@ This is a [level 2](https://github.com/Ada-Developers-Academy/pedagogy/blob/mast
...
 Each `Trip` has:
 
 **Attribute**|**Description**
 -----|-----
 id|The Trip's ID number
-passenger|The passenger on the trip
-driver|The driver for the trip
+User|The User on the trip
 start_time|When did this trip begin?
 end_time|When did this trip finish?
```
Comment (droberts-ada) [inline README.md]: Attribute `user` should probably be lower-case

File: README.md [hunk 6]
```diff
@@ -8,101 +8,79 @@ This is a [level 2](https://github.com/Ada-Developers-Academy/pedagogy/blob/mast
...
+rating|The rating given by the User, a number 1-5
 cost|How much did the passenger pay?
 
 Each `Trip` instance is able to:
 
 **Method**|**Description**
 -----|-----
-driver|retrieve the associated driver instance
-passenger|retrieve the associated passenger instance
+User|retrieve the associated User instance
```
Comment (droberts-ada) [inline README.md]: I recognize that this table existed before, but this is confusing to me. Why is `user` a method (presumably a reader) but the other attributes are not?

File: README.md [hunk 7]
```diff
@@ -8,101 +8,79 @@ This is a [level 2](https://github.com/Ada-Developers-Academy/pedagogy/blob/mast
...
 
 The `TripDispatcher` has the following responsibilities:
--  load collections of `Driver`s, `Passenger`s, and `Trip`s from CSV files
--  store and manage this data into separate collections
+-   load collections of `User`s, and `Trip`s from CSV files
+-   store and manage this data into separate collections
 
 The `TripDispatcher` does the following:
-- on instantiation, loads and creates `Trip`s, `Passenger`s, and `Driver`s and stores them into collections
+-   on instantiation, loads and creates `Trip`s, `User`s, and stores them into collections
```
Comment (droberts-ada) [inline README.md]: `Trip`s and `User`s (missing an "and")

File: README.md [hunk 8]
```diff
@@ -8,101 +8,79 @@ This is a [level 2](https://github.com/Ada-Developers-Academy/pedagogy/blob/mast
...
 The `TripDispatcher` instance is able to:
 
 **Methods**|**Description**
 -----|-----
-drivers, passengers, trips  |  retrieve the collection of `Trip`s, `Passenger`s, and `Driver`s
-find_driver  |  find an instance of `Driver` given an ID
-find_passenger  |  find an instance of `Passenger` given an ID
+passengers|retrieve the collection of `User`s who are being transported
+trips|retrieve the collection of `Trip`s  |
+find_user  |  find an instance of `User` given an ID
```
Comment (droberts-ada) [inline README.md]: Should line 78 be `users`?

File: README.md [hunk 9]
```diff
@@ -136,12 +114,12 @@ First, come up with a "plan of action" for how you want to work as a pair. Discu
...
 To start this project, take some time to get familiar with the code. Do the following in this order:
 1. Read through all of the tests
-1. Look at the provided CSV files: `support/drivers.csv`, `support/passengers.csv`, `support/trips.csv`
+1. Look at the provided CSV files: `support/drivers.csv`, `support/users.csv`, `support/trips.csv`
 1. Then look through the ruby files in the `lib` folder
 
 Create a diagram that describes how each of these classes and methods (messages) interact with one another as well as with the CSV files.
 
-**Exercise:** Look at this requirement in Wave 1: "For a given driver, calculate their total revenue for all trips. Each driver gets 80% of the trip cost after a fee of $1.65 is subtracted." Spend some time writing pseudocode for this.
+**Exercise:** Look at this requirement in Wave 1: "For a given user, calculate their total expenditure for all trips. Spend some time writing pseudocode for this.
```
Comment (droberts-ada) [inline README.md]: Missing an end-quote

File: README.md [hunk 10]
```diff
@@ -156,25 +134,74 @@ Currently our implementation saves the start and end time of each trip as a stri
...
-Our program needs a way to make new trips and appropriately assign a driver and passenger.
+Our program needs a data type to represent Drivers in our service.
+
+We will do this by creating a `Driver` class which inherits from `User`.  A `Driver` will add the following data attributes:
+
+**Attribute**|**Description**
+-----|-----
+vehicle\_identification|The driver's Vehicle Identification Number (VIN Number), Each vehicle identification number should be a specific length to ensure it is a valid vehicle identification number
+driven_trips | A list of trips the user has acted as a driver for.
+status|Indicating availability, a driver's availability should be either `:AVAILABLE` or `:UNAVAILABLE`
```
Comment (droberts-ada) [inline README.md]: specific length -> how long?

File: README.md [hunk 11]
```diff
@@ -156,25 +134,74 @@ Currently our implementation saves the start and end time of each trip as a stri
...
+
+**Create tests** to ensure that a `Driver` instance can be created successfully.
+
```
Comment (droberts-ada) [inline README.md]: They should also have a test about the status - what happens if a bad status is passed?

File: README.md [hunk 12]
```diff
@@ -156,25 +134,74 @@ Currently our implementation saves the start and end time of each trip as a stri
...
+**Method**|**Description**
+-----|-----
+load_drivers|Load the Drivers from the `support/drivers.csv` file and return a collection of `Driver` instances
+find_driver |This method takes an `id` number and returns the corresponding `Driver` instance.
+load_trips|This method should be updated to add a corresponding `Driver` to the trip instance.
+
+#### Driver methods
+
+After trips have a reference to their drivers and TripDispatcher can load a list of drivers, add the following functionality.
+
```
Comment (droberts-ada) [inline README.md]: Could you adjust this to

```markdown
After each trip has a reference to its `Driver` and TripDispatcher can load a list of `Driver`s, add the following functionality to the `Driver` class:
```

File: README.md [hunk 13]
```diff
@@ -156,25 +134,74 @@ Currently our implementation saves the start and end time of each trip as a stri
...
+**Method**|**Description**
+-----|-----
+average_rating  |  This method sums up the ratings from all a Driver's trips and returns the average
+add_trip  |  This method adds a trip to the driver's collection of trips
+total_revenue  |  This method calculates that driver's total revenue across all their trips. Each driver gets 80% of the trip cost after a fee of $1.65 is subtracted.
+net_expenditures|This method will **override** the cooresponding method in `User` and take the total amount a driver has spent as a passenger and subtract the amount they have earned as a driver (see above).  If the number is negative the driver will earn money.
```
Comment (droberts-ada) [inline README.md]: Can you make it explicit that the $1.65 is per-trip?

File: README.md [hunk 14]
```diff
@@ -187,30 +214,30 @@ You should use this information to:
 -   Modify this selected driver using a new helper method in `Driver`:
  -Add the new trip to the collection of trips for that `Driver`
  -Set the driver's status to `:UNAVAILABLE`
--   Modify the passenger for the trip using a new helper method in `Passenger`:
- -Add the new trip to the collection of trips for the `Passenger`
+-   Modify the user for the trip using a new helper method in `User`:
+ -Add the new trip to the collection of trips for the `User`
```
Comment (droberts-ada) [inline README.md]: The bullet spacing here does not render correctly.

**Comment by CheezItMan** [comment]
@droberts-ada @tildeee @dHelmgren 

It's ready as such, I'll merge it in the morning if no one has suggestions.

**Pull request merged**